File: server/pkg/mcp/protocol.go
// Package mcp implements a stateless MCP 2026-07-28 JSON-RPC facade over the
// Toolplane gRPC API. It exposes the tool registry and the durable task
// lifecycle through the MCP Tasks extension (io.modelcontextprotocol/tasks),
// translating JSON-RPC requests into Toolplane TasksService/ToolService calls.
//
// The wire shapes are pinned against the published 2026-07-28 core schema
// (schema/2026-07-28/schema.ts in modelcontextprotocol/modelcontextprotocol)
// and the Tasks extension schema (schema/draft/schema.ts in
// modelcontextprotocol/ext-tasks).
package mcp

import (
	"encoding/json"
	"fmt"
)

const (
	// JSONRPCVersion is the only JSON-RPC version MCP uses.
	JSONRPCVersion = "2.0"

	// ProtocolVersion is the MCP revision this facade speaks. There is no
	// initialize handshake in this revision: every request carries the
	// version in _meta and the server accepts or rejects each request
	// independently.
	ProtocolVersion = "2026-07-28"

	// TasksExtensionID identifies the MCP Tasks extension in capabilities.
	TasksExtensionID = "io.modelcontextprotocol/tasks"
)

// Reserved _meta keys defined by the 2026-07-28 core spec.
const (
	metaProtocolVersion    = "io.modelcontextprotocol/protocolVersion"
	metaClientCapabilities = "io.modelcontextprotocol/clientCapabilities"
)

// Toolplane-namespaced _meta keys. The prefix follows the spec's reverse-DNS
// naming rules for non-reserved vendor keys.
const (
	// SessionIDMetaKey lets a client bind a request to an existing Toolplane
	// session (the spec's explicit-handle guidance). When absent, the gateway
	// auto-provisions one session per API key.
	SessionIDMetaKey = "dev.toolplane/session_id"

	// LastSeqMetaKey carries the chunk cursor on tasks/get: the sequence
	// number of the first chunk the client still needs. Chunks with a
	// sequence >= last_seq are returned.
	LastSeqMetaKey = "dev.toolplane/last_seq"

	// ChunksMetaKey carries streaming chunks on tasks/get results while the
	// underlying request is still executing.
	ChunksMetaKey = "dev.toolplane/chunks"

	// TaskIDDataKey appears in error data when a synchronous tools/call times
	// out; the task keeps running and can be tracked with this ID.
	TaskIDDataKey = "toolplaneTaskId"
)

// JSON-RPC and MCP error codes.
const (
	CodeParseError                  = -32700
	CodeInvalidRequest              = -32600
	CodeMethodNotFound              = -32601
	CodeInvalidParams               = -32602
	CodeInternalError               = -32603
	CodeUnsupportedProtocolVersion  = -32022
	CodeMissingRequiredCapability   = -32021
)

// MCP Tasks extension task statuses.
const (
	TaskStatusWorking       = "working"
	TaskStatusInputRequired = "input_required"
	TaskStatusCompleted     = "completed"
	TaskStatusFailed        = "failed"
	TaskStatusCancelled     = "cancelled"
)

// Result types carried by the resultType discriminator field. Servers
// implementing 2026-07-28 MUST include resultType on every result.
const (
	ResultTypeComplete      = "complete"
	ResultTypeInputRequired = "input_required"
	ResultTypeTask          = "task"
)

// Request is a JSON-RPC 2.0 request. ID stays raw because JSON-RPC ids may be
// strings, numbers, or null.
type Request struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params,omitempty"`
}

// IsNotification reports whether the request carries no id and therefore
// expects no response.
func (r *Request) IsNotification() bool {
	return len(r.ID) == 0 || string(r.ID) == "null"
}

// Error is a JSON-RPC error object.
type Error struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
	Data    any    `json:"data,omitempty"`
}

func (e *Error) Error() string {
	return fmt.Sprintf("jsonrpc error %d: %s", e.Code, e.Message)
}

// Response is a JSON-RPC 2.0 response envelope.
type Response struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id"`
	Result  any             `json:"result,omitempty"`
	Error   *Error          `json:"error,omitempty"`
}

func newResultResponse(id json.RawMessage, result any) Response {
	return Response{JSONRPC: JSONRPCVersion, ID: id, Result: result}
}

func newErrorResponse(id json.RawMessage, err *Error) Response {
	return Response{JSONRPC: JSONRPCVersion, ID: id, Error: err}
}

func errInvalidRequest(msg string) *Error {
	return &Error{Code: CodeInvalidRequest, Message: msg}
}

func errMethodNotFound(msg string) *Error {
	return &Error{Code: CodeMethodNotFound, Message: msg}
}

func errInvalidParams(msg string) *Error {
	return &Error{Code: CodeInvalidParams, Message: msg}
}

func errInternal(msg string, data any) *Error {
	return &Error{Code: CodeInternalError, Message: msg, Data: data}
}

// unsupportedVersionError builds the 2026-07-28 UnsupportedProtocolVersionError.
func unsupportedVersionError(requested string) *Error {
	return &Error{
		Code:    CodeUnsupportedProtocolVersion,
		Message: "Unsupported protocol version",
		Data: map[string]any{
			"supported": []string{ProtocolVersion},
			"requested": requested,
		},
	}
}

// baseParams captures the fields common to every MCP request params object.
type baseParams struct {
	Meta map[string]any `json:"_meta"`
}

// requestMeta is the decoded, validated _meta of a single request.
type requestMeta struct {
	protocolVersion string
	capabilities    map[string]any
	sessionID       string
	lastSeq         int32
	hasLastSeq      bool
}

// parseRequestMeta validates the modern per-request metadata. The 2026-07-28
// spec requires protocolVersion and clientCapabilities on every request.
func parseRequestMeta(params json.RawMessage) (requestMeta, *Error) {
	var base baseParams
	if len(params) > 0 {
		if err := json.Unmarshal(params, &base); err != nil {
			return requestMeta{}, errInvalidParams("params must be a JSON object: " + err.Error())
		}
	}
	if base.Meta == nil {
		return requestMeta{}, errInvalidRequest("missing _meta: MCP 2026-07-28 is stateless and requires _meta with " + metaProtocolVersion + " and " + metaClientCapabilities + " on every request")
	}

	meta := requestMeta{}

	version, _ := base.Meta[metaProtocolVersion].(string)
	if version == "" {
		return requestMeta{}, errInvalidRequest("missing _meta." + metaProtocolVersion + ": every request must declare its MCP protocol version")
	}
	meta.protocolVersion = version
	if version != ProtocolVersion {
		return meta, unsupportedVersionError(version)
	}

	caps, _ := base.Meta[metaClientCapabilities].(map[string]any)
	if caps == nil {
		return requestMeta{}, errInvalidRequest("missing _meta." + metaClientCapabilities + ": every request must declare client capabilities (use {} for none)")
	}
	meta.capabilities = caps

	if sessionID, ok := base.Meta[SessionIDMetaKey].(string); ok {
		meta.sessionID = sessionID
	}

	switch cursor := base.Meta[LastSeqMetaKey].(type) {
	case float64:
		meta.lastSeq = int32(cursor)
		meta.hasLastSeq = true
	case int64:
		meta.lastSeq = int32(cursor)
		meta.hasLastSeq = true
	}

	return meta, nil
}

// clientSupportsTasks reports whether the client advertised the Tasks
// extension in this request's capabilities. Servers MUST NOT return a task
// handle to a client that did not declare support.
func (m requestMeta) clientSupportsTasks() bool {
	extensions, _ := m.capabilities["extensions"].(map[string]any)
	if extensions == nil {
		return false
	}
	_, ok := extensions[TasksExtensionID]
	return ok
}

File: clients/typescript-mcp-adapter/src/protocol.ts
import { McpError } from '@modelcontextprotocol/sdk/types.js';

/**
 * MCP 2026-07-28 stateless-core wire contract shared by the adapter's request
 * handlers. Mirrors server/pkg/mcp/protocol.go: per-request _meta carries the
 * protocol version and client capabilities (no initialize handshake), every
 * result carries a resultType discriminator, and the Tasks extension
 * (io.modelcontextprotocol/tasks) maps Toolplane's durable requests to
 * pollable task handles.
 */

export const PROTOCOL_VERSION = '2026-07-28';

export const TASKS_EXTENSION_ID = 'io.modelcontextprotocol/tasks';

export const META_PROTOCOL_VERSION = 'io.modelcontextprotocol/protocolVersion';

export const META_CLIENT_CAPABILITIES = 'io.modelcontextprotocol/clientCapabilities';

/** Chunk cursor sent by clients on tasks/get: "first chunk seq I still need". */
export const LAST_SEQ_META_KEY = 'dev.toolplane/last_seq';

/** Server-to-client chunk window carried in tasks/get _meta while working. */
export const CHUNKS_META_KEY = 'dev.toolplane/chunks';

/** Error data key carrying the Toolplane task (request) id. */
export const TASK_ID_DATA_KEY = 'toolplaneTaskId';

export const RESULT_TYPE_COMPLETE = 'complete';

export const RESULT_TYPE_TASK = 'task';

export const TASK_STATUS_WORKING = 'working';

export const TASK_STATUS_COMPLETED = 'completed';

export const TASK_STATUS_FAILED = 'failed';

export const TASK_STATUS_CANCELLED = 'cancelled';

/** Polling interval suggested to MCP clients in every task payload. */
export const TASK_POLL_INTERVAL_MS = 500;

export const ERROR_INVALID_REQUEST = -32600;

export const ERROR_METHOD_NOT_FOUND = -32601;

export const ERROR_INVALID_PARAMS = -32602;

export const ERROR_INTERNAL = -32603;

export const ERROR_UNSUPPORTED_PROTOCOL_VERSION = -32022;

/**
 * Exact error marker CancelRequest records on a cancelled Toolplane request;
 * distinguishes user cancellation from other failures (which surface as the
 * generic failed task status). Mirrors requestCancelledError in tasks.go.
 */
export const REQUEST_CANCELLED_ERROR = 'Request was cancelled';

export type TaskStatus =
  | typeof TASK_STATUS_WORKING
  | typeof TASK_STATUS_COMPLETED
  | typeof TASK_STATUS_FAILED
  | typeof TASK_STATUS_CANCELLED;

/**
 * Wire shape shared by the flat CreateTaskResult returned from tools/call
 * (resultType "task") and the DetailedTask variants returned by tasks/get
 * (resultType "complete"). Field names follow the Tasks extension schema.
 * The task identifier is a Toolplane request id: requests are the durable
 * units that provider machines claim and execute.
 */
export interface TaskPayload {
  resultType: string;
  taskId: string;
  status: TaskStatus;
  statusMessage?: string;
  createdAt: string;
  lastUpdatedAt: string;
  ttlMs: number | null;
  pollIntervalMs?: number;
  result?: Record<string, unknown>;
  error?: { code: number; message: string; data?: unknown };
  _meta?: Record<string, unknown>;
  [key: string]: unknown;
}

/** Retained chunk window for a working task, sliced at the client's cursor. */
export interface ChunkWindow {
  requestId: string;
  startSeq: number;
  nextSeq: number;
  chunks: unknown[];
}

/** Decoded, validated _meta of a single 2026 request. */
export interface RequestMeta {
  capabilities: Record<string, unknown>;
  lastSeq?: number;
}

export function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/**
 * Reports whether the params object carries a 2026 protocol version in _meta,
 * which switches a tools/list or tools/call handler into the stateless surface
 * (resultType discriminators, task handles). Params without it are served with
 * legacy semantics negotiated at initialize.
 */
export function has2026Meta(params: unknown): boolean {
  if (!isRecord(params) || !isRecord(params._meta)) {
    return false;
  }

  return META_PROTOCOL_VERSION in params._meta;
}

/**
 * Validates the stateless per-request metadata. The 2026-07-28 spec requires
 * protocolVersion and clientCapabilities on every request. Mirrors
 * parseRequestMeta in server/pkg/mcp/protocol.go.
 */
export function parseRequestMeta(params: unknown): RequestMeta {
  const meta = isRecord(params) && isRecord(params._meta) ? params._meta : undefined;
  if (!meta) {
    throw new McpError(
      ERROR_INVALID_REQUEST,
      `missing _meta: MCP 2026-07-28 is stateless and requires _meta with ${META_PROTOCOL_VERSION} and ${META_CLIENT_CAPABILITIES} on every request`,
    );
  }

  const version = typeof meta[META_PROTOCOL_VERSION] === 'string' ? meta[META_PROTOCOL_VERSION] : '';
  if (!version) {
    throw new McpError(
      ERROR_INVALID_REQUEST,
      `missing _meta.${META_PROTOCOL_VERSION}: every request must declare its MCP protocol version`,
    );
  }
  if (version !== PROTOCOL_VERSION) {
    throw new McpError(ERROR_UNSUPPORTED_PROTOCOL_VERSION, 'Unsupported protocol version', {
      supported: [PROTOCOL_VERSION],
      requested: version,
    });
  }

  const capabilities = meta[META_CLIENT_CAPABILITIES];
  if (!isRecord(capabilities)) {
    throw new McpError(
      ERROR_INVALID_REQUEST,
      `missing _meta.${META_CLIENT_CAPABILITIES}: every request must declare client capabilities (use {} for none)`,
    );
  }

  let lastSeq: number | undefined;
  if (LAST_SEQ_META_KEY in meta) {
    const cursor = meta[LAST_SEQ_META_KEY];
    if (typeof cursor !== 'number' || !Number.isInteger(cursor) || cursor < 0 || cursor > 2147483647) {
      throw new McpError(ERROR_INVALID_PARAMS, `_meta.${LAST_SEQ_META_KEY} must be a non-negative 32-bit integer`);
    }
    lastSeq = cursor;
  }

  return { capabilities, lastSeq };
}

/**
 * Reports whether the client advertised the Tasks extension in this request's
 * capabilities. Servers MUST NOT return a task handle to a client that did not
 * declare support. Mirrors requestMeta.clientSupportsTasks.
 */
export function clientSupportsTasks(capabilities: Record<string, unknown>): boolean {
  const extensions = capabilities.extensions;
  return isRecord(extensions) && TASKS_EXTENSION_ID in extensions;
}

export function isTerminalTaskStatus(status: TaskStatus): boolean {
  return (
    status === TASK_STATUS_COMPLETED ||
    status === TASK_STATUS_FAILED ||
    status === TASK_STATUS_CANCELLED
  );
}
